### DISCUSSION TITLE
How can I run an asynchronous function in a Custom Link? (resolve race condition with refresh token of third-party provider)

### DISCUSSION THREAD

**Discussion by herzog0** [trpc/trpc#3470] [General]
Hi! 
I'm having a very specific problem in which I use a third-party authentication provider (Clerk) that invalidates and refreshes their token in the browser very frequently (1-2 minutes).  But Clerk is only able to run js and refresh the token if my page has focus. So what happens, very frequently, is that a user change focus to another tab or software, then when they return to my page a race condition occurs, where Clerk begins to refresh the token but, simultaneously, tRPC is already making a request to my server.  So, by the time this first tRPC's request arrives at my server, Clerk hasn't had the time to refresh it yet, and users get an unauthorized error.  

This is not a major problem right now because of automatic retries, but I'm trying to solve it anyway.  

My logic was to add a Custom Link to tRPC, requesting the Clerk authed user and awaiting for it to refresh (if needed) before returning, but apparently I can't run asynchronous functions inside its body.   

Could someone shine a light on how to resolve this, please?

Thank you!!

**Comment by KATT** [comment]
Links use observables that are like promises on steroids, but they are unfortunately a bit harder to write

Look at the built-in ones to see how we do async stuff (esp dedupe one or the cache one)

Edit: I forgot to link here! https://github.com/trpc/trpc/tree/main/packages/client/src/links

**Reply by herzog0** [reply]
Thanks!

**Reply by osamaqarem** [reply]
been trying to do this today as well. this seems to work but can't promise no issues. unsure I fully understand the observables.

```ts
const authRefreshLink: TRPCLink<AppRouter> = () => {
  let authRefresh: Nullable<ReturnType<AuthStoreActions["maybeRefresh"]>> = null

  return ({ next, op }) => {
    return observable(observer => {
      const unsubscribe = next(op).subscribe({
        next(value) {
          observer.next(value)
        },
        complete() {
          observer.complete()
        },
        async error(err) {
          if (err.message !== ApiError.InvalidAccessToken) {
            observer.error(err)
          } else {
            authRefresh ??= useAuth.getState().actions.maybeRefresh()
            const outcome = await authRefresh

            if (outcome === "authorized") {
              const retry = next(op).subscribe({
                next(value) {
                  authRefresh = null
                  observer.next(value)
                  retry.unsubscribe()
                },
                error(error) {
                  authRefresh = null
                  observer.error(err)
                  retry.unsubscribe()
                },
                complete() {
                  authRefresh = null
                  observer.complete()
                  retry.unsubscribe()
                },
              })
            } else {
              authRefresh = null
              observer.error(err)
            }
          }
        },
      })

      return unsubscribe
    })
  }
}
```
